bindings/c/test/unit/unit_tests.cpp

```diff
@@ -0,0 +1,1841 @@
+
+    // Callbacks will be triggered with operation_cancelled errors once the
+    // too_many_watches error fires, as the other futures will go out of scope
+    // and be cleaned up. The future which too_many_watches occurs on is
+    // nondeterministic, so each future is checked.
+    fdb_check(f1.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+    fdb_check(f2.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+    fdb_check(f3.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+    fdb_check(f4.set_callback(
+        +[](FDBFuture *f, void *param) {
+          fdb_error_t err = fdb_future_get_error(f);
+          if (err != 1101) { // operation_cancelled
+            CHECK(err == 1032); // too_many_watches
+          }
+          auto *event = static_cast<std::shared_ptr<FdbEvent> *>(param);
+          (*event)->set();
+          delete event;
+        }, new std::shared_ptr<FdbEvent>(event)));
+
+    event->wait();
+    break;
+  }
+
+  // Reset available number of watches.
+  max_watches = 10000;
+  fdb_check(fdb_database_set_option(db, FDB_DB_OPTION_MAX_WATCHES,
+                                    (const uint8_t *)&max_watches, 8));
+}
+
+TEST_CASE("fdb_transaction_watch") {
+  fdb::Transaction tr(db);
+
+  struct Context {
+    FdbEvent event;
+  };
+  Context context;
+
+  while (1) {
+    fdb::EmptyFuture f1 = tr.watch(key("foo"));
+    fdb::EmptyFuture f2 = tr.commit();
+
+    fdb_error_t err = wait_future(f2);
+    if (err) {
+      fdb::EmptyFuture f3 = tr.on_error(err);
+      fdb_check(wait_future(f3));
+      continue;
+    }
+
+    fdb_check(f1.set_callback(
+        +[](FDBFuture *, void *param) {
+          auto *context = static_cast<Context *>(param);
+          context->event.set();
+        },
+        &context));
+    break;
+  }
+
+  // Insert data for key "foo" to trigger the watch.
+  insert_data(db, create_data({ { "foo", "bar" } }));
+  context.event.wait();
```

Comment:
This is only firing because f1 is getting destroyed. Let's set "foo" to something that's not "bar" before creating the watch, extend the lifetime of f1, and assert that f1 completes successfully in the callback